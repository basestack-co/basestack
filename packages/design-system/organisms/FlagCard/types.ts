import { PositionProps, SpaceProps } from "styled-system";
import { PopupItems } from "../../molecules/Popup";

export interface FlagCardProps extends SpaceProps, PositionProps {
  /**
   * Card title
   */
  title: string;
  /**
   * Card title
   */
  description: string;
  /**
   * Card title
   */
  date: string;
  /**
   * Environments where flag is created
   */
  environments: Array<{
    text: string;
    isFlagOn: boolean;
  }>;
  /**
   * Popup items array
   */
  popupItems: Array<PopupItems>;
}
