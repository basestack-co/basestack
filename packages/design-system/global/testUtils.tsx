import { type RenderOptions, render } from "@testing-library/react";
import type React from "react";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "../theme/lightTheme";

interface Props {
  children: React.ReactNode;
}

const WithThemeProvider = ({ children }: Props) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithThemeProvider, ...options });

export { renderWithTheme };
