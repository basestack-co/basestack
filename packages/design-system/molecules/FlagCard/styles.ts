import styled from "styled-components";
import { Label, Card } from "../../atoms";

export const StyledCard = styled(Card)`
  transition: box-shadow 0.2s ease-in-out;
  &:hover {
    box-shadow: ${({ theme }) => theme.shadow.elevation4};
  }
`;

export const Labels = styled.div`
  display: flex;
  flex-wrap: wrap;
  margin-bottom: ${({ theme }) => theme.spacing.s3};
`;

export const StyledLabel = styled(Label)`
  margin-right: ${({ theme }) => theme.spacing.s1};
  margin-bottom: ${({ theme }) => theme.spacing.s1};
`;
