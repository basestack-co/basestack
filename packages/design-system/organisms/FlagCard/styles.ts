import styled from "styled-components";
import { space, position, compose } from "styled-system";
import { Label, Card } from "../../atoms";

export const StyledCard = styled(Card)`
  ${compose(space, position)};
  position: relative;
`;

export const Labels = styled.div`
  display: flex;
  flex-wrap: wrap;
  margin-bottom: ${({ theme }) => theme.spacing.s3};
`;

export const StyledLabel = styled(Label)`
  margin-right: ${({ theme }) => theme.spacing.s1};
  margin-bottom: ${({ theme }) => theme.spacing.s1};
`;

export const PopupWrapper = styled.div`
  height: 0;
`;
