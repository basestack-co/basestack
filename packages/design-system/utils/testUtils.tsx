import defaultTheme from "@basestack/design-system/theme/lightTheme";
import isPropValid from "@emotion/is-prop-valid";
import { RenderResult, render } from "@testing-library/react";
import React from "react";
import { StyleSheetManager, ThemeProvider } from "styled-components";

const Provider = ({
  children,
  theme = defaultTheme,
}: {
  children: React.ReactNode;
  theme: any;
}) => (
  <ThemeProvider theme={theme}>
    <StyleSheetManager shouldForwardProp={isPropValid}>
      {children}
    </StyleSheetManager>
  </ThemeProvider>
);

export const renderWithTheme = (
  ui: React.ReactElement,
  options?: any,
): RenderResult => render(ui, { wrapper: Provider, ...options });
