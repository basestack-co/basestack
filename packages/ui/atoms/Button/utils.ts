import { Size, Target, Type, Variant } from "@/components/atoms/Button/types";
import { css, DefaultTheme } from "styled-components";
import { darken, rem, transparentize } from "polished";

export const handleButtonSize = (size?: Size) => {
  switch (size) {
    case "large":
      return css`
        height: ${rem("60px")};
        font-size: ${rem("18px")};
        line-height: ${rem("25px")};
        padding: 0 ${rem("40px")};
      `;
    case "medium":
      return css`
        height: ${rem("54px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("34px")};
      `;
    case "normal":
      return css`
        height: ${rem("48px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("30px")};
      `;
    default:
    case "small":
      return css`
        height: ${rem("46px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("24px")};
      `;
    case "xSmall":
      return css`
        height: ${rem("42px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("20px")};
      `;
  }
};

export const handleButtonsColors = (theme: DefaultTheme) => ({
  primary: {
    outline: {
      color: theme.colors.primary,
      borderColor: theme.colors.primary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.primary,
    },
  },
  secondary: {
    outline: {
      color: theme.colors.secondary,
      borderColor: theme.colors.secondary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.secondary,
    },
  },
  tertiary: {
    outline: {
      color: theme.colors.gray3,
      borderColor: theme.colors.gray3,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: theme.colors.gray6,
    },
  },
  neutral: {
    outline: {
      color: theme.colors.gray2,
      borderColor: theme.colors.gray2,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: "transparent",
    },
  },
  success: {
    outline: {
      color: theme.colors.success,
      borderColor: theme.colors.success,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.success,
    },
  },
  warning: {
    outline: {
      color: theme.colors.warning,
      borderColor: theme.colors.warning,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.warning,
    },
  },
  danger: {
    outline: {
      color: theme.colors.error,
      borderColor: theme.colors.error,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.error,
    },
  },
});

export const darkenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return darken(
    amount || 0.1,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};

export const lightenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return transparentize(
    amount || 0.6,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};
