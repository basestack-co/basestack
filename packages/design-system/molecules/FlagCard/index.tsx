import React, { memo, forwardRef } from "react";
import { SpaceProps } from "styled-system";
import { useTheme } from "styled-components";
import { Text } from "../../atoms";
import { Labels, StyledCard, StyledLabel } from "./styles";

export interface FlagCardProps extends SpaceProps {
  /**
   * Card title
   */
  title: string;
  /**
   * Card title
   */
  description: string;
  /**
   * Card title
   */
  date: string;
  /**
   * Environments where flag is created
   */
  environments: Array<{
    text: string;
    isFlagOn: boolean;
  }>;
}

const FlagCard = forwardRef<HTMLDivElement, FlagCardProps>(
  ({ title, description, environments, date, ...props }, ref) => {
    const theme = useTheme();

    return (
      <StyledCard ref={ref} testId="flag-card" p={theme.spacing.s5}>
        <Text data-testid="flag-title" size="large" mb={theme.spacing.s2}>
          {title}
        </Text>
        <Text data-testid="flag-description" size="small" mb={theme.spacing.s3}>
          {description}
        </Text>
        <Labels data-testid="flag-labels">
          {environments.map((environment, index) => {
            return (
              !!environment.text && (
                <StyledLabel
                  key={index.toString()}
                  testId={`${environment.text}-flag-label`}
                  text={environment.text}
                  variant={environment.isFlagOn ? "success" : "default"}
                />
              )
            );
          })}
        </Labels>
        <Text data-testid="flag-date" mt="auto" size="small" muted>
          {date}
        </Text>
      </StyledCard>
    );
  }
);

FlagCard.displayName = "FlagCard";

export default memo(FlagCard);
