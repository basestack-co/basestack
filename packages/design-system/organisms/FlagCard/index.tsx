import React, { memo, forwardRef, useState, useCallback, useRef } from "react";
import { useFloating, autoUpdate } from "@floating-ui/react-dom";
import { useClickAway } from "@basestack/hooks";
import { useTheme } from "styled-components";
import { useTransition, animated, config } from "react-spring";
import { IconButton, Text } from "../../atoms";
import { Popup } from "../../molecules";
import { Labels, StyledCard, StyledLabel, PopupWrapper } from "./styles";
import { FlagCardProps } from "./types";
import { scaleInTopRight } from "../../animations/springs";

const AnimatedPopup = animated(Popup);

const FlagCard = forwardRef<HTMLDivElement, FlagCardProps>(
  ({ title, description, environments, date, popupItems, ...props }, ref) => {
    const theme = useTheme();
    const popupWrapperRef = useRef(null);
    const [isPopupOpen, setIsPopupOpen] = useState(false);
    const { x, y, strategy, refs } = useFloating({
      placement: "bottom-end",
      whileElementsMounted: autoUpdate,
    });

    const onClickMore = useCallback(() => {
      setIsPopupOpen((prevState) => !prevState);
    }, []);

    const transitionPopup = useTransition(isPopupOpen, {
      config: { ...config.default, duration: 150 },
      ...scaleInTopRight,
    });

    useClickAway(popupWrapperRef, () => {
      setIsPopupOpen(false);
    });

    return (
      <StyledCard
        ref={ref}
        testId="flag-card"
        hasHoverAnimation
        p={theme.spacing.s5}
        {...props}
      >
        <Text data-testid="flag-title" size="large" mb={theme.spacing.s2}>
          {title}
        </Text>
        <Text data-testid="flag-description" size="small" mb={theme.spacing.s3}>
          {description}
        </Text>
        <Labels data-testid="flag-labels">
          {environments.map((environment) => {
            return (
              !!environment.name && (
                <StyledLabel
                  key={environment.id}
                  testId={`${environment.name}-flag-label`}
                  text={environment.name}
                  variant={environment.enabled ? "success" : "default"}
                />
              )
            );
          })}
        </Labels>
        <Text data-testid="flag-date" mt="auto" size="small" muted>
          {date}
        </Text>
        <PopupWrapper ref={popupWrapperRef}>
          <IconButton
            ref={refs.setReference}
            position="absolute"
            top="14px"
            right="14px"
            icon="more_horiz"
            onClick={onClickMore}
          />
          {transitionPopup(
            (styles, item) =>
              item &&
              popupItems.length > 0 && (
                <AnimatedPopup
                  style={styles}
                  ref={refs.setFloating}
                  position={strategy}
                  top={y}
                  left={x}
                  items={popupItems}
                />
              )
          )}
        </PopupWrapper>
      </StyledCard>
    );
  }
);

FlagCard.displayName = "FlagCard";

export default memo(FlagCard);
