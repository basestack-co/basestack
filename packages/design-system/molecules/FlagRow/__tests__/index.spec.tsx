import React from "react";
import { cleanup } from "@testing-library/react";
import { renderWithTheme } from "../../../utils/testUtils";
import FlagCard from "..";

describe("FlagCard Molecule tests", () => {
  afterEach(cleanup);

  test("should render FlagCard correctly", () => {
    const { asFragment } = renderWithTheme(
      <FlagCard
        onClickMore={jest.fn()}
        title="header_size"
        description="Display new header for only a number of users"
        environments={[
          { text: "Development", isFlagOn: true },
          { text: "Staging", isFlagOn: false },
          { text: "Production", isFlagOn: false },
        ]}
        date="Created 11 jan 2022"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render FlagCard with core elements", () => {
    const { getByTestId, getByText } = renderWithTheme(
      <FlagCard
        onClickMore={jest.fn()}
        title="header_size"
        description="Display new header for only a number of users"
        environments={[
          { text: "Development", isFlagOn: true },
          { text: "Staging", isFlagOn: false },
          { text: "Production", isFlagOn: false },
        ]}
        date="Created 11 jan 2022"
      />
    );
    const title = getByTestId("flag-title");
    const description = getByTestId("flag-description");
    const labels = getByTestId("flag-labels");
    const devLabel = getByTestId("Development-flag-label");
    const stagLabel = getByTestId("Staging-flag-label");
    const prodLabel = getByTestId("Production-flag-label");
    const date = getByTestId("flag-date");

    expect(title).toBeVisible();
    expect(description).toBeVisible();
    expect(labels).toBeVisible();
    expect(devLabel).toBeVisible();
    expect(stagLabel).toBeVisible();
    expect(prodLabel).toBeVisible();
    expect(date).toBeVisible();
  });
});
