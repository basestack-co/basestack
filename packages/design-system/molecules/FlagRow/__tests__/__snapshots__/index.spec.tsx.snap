// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlagCard Molecule tests should render FlagCard correctly 1`] = `
<DocumentFragment>
  .c3 {
  font-size: 0.875rem;
  line-height: 1.25rem;
  font-weight: 400;
  margin-bottom: 0.75rem;
  font-family: 'Roboto',sans-serif;
  color: #000000;
}

.c8 {
  font-size: 0.875rem;
  line-height: 1.25rem;
  font-weight: 400;
  margin-top: auto;
  font-family: 'Roboto',sans-serif;
  color: #545454;
}

.c2 {
  font-size: 1.125rem;
  line-height: 1.5rem;
  font-weight: 500;
  margin-bottom: 0.5rem;
  font-family: 'Roboto',sans-serif;
  color: #000000;
}

.c0 {
  padding: 1.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  box-shadow: rgba(0,0,0,0.1) 0px 1px 3px 0px,rgba(0,0,0,0.06) 0px 1px 2px 0px;
  background-color: #FFFFFF;
  border-radius: 0.375rem;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0.25rem 0.375rem;
  border-radius: 4px;
  font-size: 0.875rem;
  line-height: 1rem;
  font-weight: 400;
  color: #FFFFFF;
  background-color: #05944F;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0.25rem 0.375rem;
  border-radius: 4px;
  font-size: 0.875rem;
  line-height: 1rem;
  font-weight: 400;
  color: #545454;
  background-color: #E2E2E2;
}

.c1 {
  -webkit-transition: box-shadow 0.2s ease-in-out;
  transition: box-shadow 0.2s ease-in-out;
}

.c1:hover {
  box-shadow: rgba(0,0,0,0.1) 0px 10px 15px -3px,rgba(0,0,0,0.05) 0px 4px 6px -2px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-bottom: 0.75rem;
}

.c6 {
  margin-right: 0.25rem;
  margin-bottom: 0.25rem;
}

<div
    class="c0 c1"
    data-testid="flag-card"
  >
    <h3
      class="c2"
      data-testid="flag-title"
    >
      header_size
    </h3>
    <p
      class="c3"
      data-testid="flag-description"
    >
      Display new header for only a number of users
    </p>
    <div
      class="c4"
      data-testid="flag-labels"
    >
      <div
        class="c5 c6"
        data-testid="Development-flag-label"
      >
        Development
      </div>
      <div
        class="c7 c6"
        data-testid="Staging-flag-label"
      >
        Staging
      </div>
      <div
        class="c7 c6"
        data-testid="Production-flag-label"
      >
        Production
      </div>
    </div>
    <p
      class="c8"
      data-testid="flag-date"
    >
      Created 11 jan 2022
    </p>
  </div>
</DocumentFragment>
`;
